### PR TITLE
SE and ASR joint training

### PR DESCRIPTION
**Pull request by LiChenda**
After some debugging and tuning, the current refactored joint training branch can be trained to convergence normally.
I'm still working on it for code cleanness and quality, aiming to merge the `joint-training` function to the master soon.

### CODE REVIEW

File: egs2/TEMPLATE/asr1/scripts/utils/show_asr_result.sh [hunk 1]
```diff
@@ -44,17 +44,17 @@ cat << EOF
 EOF
 
 while IFS= read -r expdir; do
-    if ls "${expdir}"/*/*/score_*/result.txt &> /dev/null; then
+    if ls "${expdir}"/*/score_*/result.txt &> /dev/null; then
         echo "## $(basename ${expdir})"
         for type in wer cer ter; do
-            if ls "${expdir}"/*/*/score_${type}/result.txt &> /dev/null; then
+            if ls "${expdir}"/*/score_${type}/result.txt &> /dev/null; then
```
Comment (LiChenda) [inline egs2/TEMPLATE/asr1/scripts/utils/show_asr_result.sh:50]: The previous `/*/*/` will failed when use `exp` parameter to specify a directory. I removed a `*`, now it look ok.

Comment (kamo-naoyuki) [reply egs2/TEMPLATE/asr1/scripts/utils/show_asr_result.sh:50]: I did modify asr.sh long time ago to work with */* for readability. Please keep as it is and should fix enh_asr.sh rather.

https://github.com/espnet/espnet/blob/e42b1ef56b433266e2ed51e87d7060982ae7d27a/egs2/TEMPLATE/asr1/asr.sh#L1185

Comment (LiChenda) [reply egs2/TEMPLATE/asr1/scripts/utils/show_asr_result.sh:50]: Before this modification, I have also referred to `asr.sh`, and run `./scripts/utils/show_asr_result.sh exp/some_exp_dir`, but found it not work in asr task either.  After remove a `*`, then both tasks works fine. I'm not sure if it is something else wrong in my environment. Could you please try it in the asr task?

Comment (kamo-naoyuki) [reply egs2/TEMPLATE/asr1/scripts/utils/show_asr_result.sh:50]: It works fine for me. If the current code doesn't work really, the current ci should also fail, so something is wrong in your side.

Please check asr.sh again. 

`${asr_exp}/${inference_tag}/${dset}/score_${_type}` is match with `"${expdir}"/*/*/score_${type}`.

Comment (kamo-naoyuki) [reply egs2/TEMPLATE/asr1/scripts/utils/show_asr_result.sh:50]: The expected directory structure is as following:

```
- exp
   -  asr_train*
       -  inference*
           -  eval_set1
               -  score_cer
               -  score_wer
           -  eval_set2
               -  score_cer
               -  score_wer
           - ...

```

Comment (LiChenda) [reply egs2/TEMPLATE/asr1/scripts/utils/show_asr_result.sh:50]: Thank you @kamo-naoyuki ! I find the problem. The ${inference_tag} is missing in my code base, thus the directory structure is one-level shallower. I'll fix this, Thank you!

**Comment by LiChenda** [comment]
I've finished the tests. Now the coverage is okey. Could you help us for the code review ?

**Comment by sw005320** [comment]
I’ll review it, but could @Emrys365 and @shincling self-review this PR, as this PR is very large?
@neillu23, could you also help us to review this PR, when you have time?

**Comment by neillu23** [comment]
No problem, I will check this PR.

> I’ll review it, but could @Emrys365 and @shincling self-review this PR, as this PR is very large?
> @neillu23, could you also help us to review this PR, when you have time?

**Comment by shincling** [comment]
> I’ll review it, but could @Emrys365 and @shincling self-review this PR, as this PR is very large?
> @neillu23, could you also help us to review this PR, when you have time?

OK. Wangyou and me are already testing some of them. Will check it thoroughly.

**Comment by LiChenda** [comment]
Another [PR](https://github.com/espnet/espnet/pull/4226) provides the same feature but a better design. So I'll close this one.

**Pull request closed without merging**